sycl/doc/Array_Kernel_Parameters.md

```diff
@@ -0,0 +1,435 @@
+};
+```
+
+Each entry in the kernel_signatures table contains three values:
+1) an encoding of the type of capture object member
+2) a field that encodes additional properties, and
+3) an offset within a block of memory where the value of that
+4) kernel argument is placed.
+
+The previous sections described how kernel arguments are handled today.
+The next three sections describe support for arrays.
+
+<h3>Fix 1: Kernel Arguments that are Standard-Layout Arrays</h3>
+
+As described earlier, each variable captured by a lambda that comprises a
+SYCL kernel becomes a parameter of the kernel caller function.
+For arrays, simply allowing them through would result in a
+function parameter of array type. This is not supported in C++.
+Therefore, the array needing capture is wrapped in a struct for
+the purposes of passing to the device. Once received on the device
+within its wrapper, the array is copied into the local capture object.
+All references to the array within the kernel body are directed to
+the non-wrapped array which is a member of the local capture object.
+
+<h4>Source code fragment:</h4>
+
+```C++
+  int array[100];
+  auto outBuf = buffer<int, 1>(&output[0], num_items);
+
+  myQueue.submit([&](handler &cgh) {
+    auto outAcc = outBuf.get_access<access::mode::write>(cgh);
+    cgh.parallel_for<class Worker>(num_items, [=](cl::sycl::id<1> index) {
+      outAcc[index] = array[index.get(0)];
+    });
+  });
+```
+
+<h4>Integration header produced:</h4>
+
+```C++
+static constexpr
+const kernel_param_desc_t kernel_signatures[] = {
+  //--- _ZTSZZ4mainENKUlRN2cl4sycl7handlerEE16->18clES2_E6Worker
+  { kernel_param_kind_t::kind_accessor, 4062, 0 },
+  { kernel_param_kind_t::kind_std_layout, 400, 32 },
+};
+```
+
+<h4>The changes to device code made to support this extension, in pseudo-code:</h4>
+
+```C++
+struct Capture {
+    Accessor outAcc;
+    int array[100];
+    () {
+        // Body
+    }
+}
+
+struct wrapper {
+    int array[100];
+};
+spir_kernel void caller(
+    __global int* AccData, // arg1 of Accessor init function
+    range<1> AccR1,        // arg2 of Accessor init function
+    range<1> AccR2,        // arg3 of Accessor init function
+    id<1> I,               // arg4 of Accessor init function
+    struct wrapper w_s     // Pass the array wrapped in a struct
+)
+{
+    // Local capture object
+    struct Capture local;
+
+    // Reassemble capture object from parts
+    // Initialize array using existing clang Initialization mechanisms
+    local.array = w_s; 
+    // Call accessor�s init function
+    Accessor::init(&local.outAcc, AccData, AccR1, AccR2, I);
+
+    callee(&local, id<1> wi);
+}
+```
+
+The sharp-eyed reviewer of `SemaSYCL.cpp` will notice that the array
+is actually double-wrapped in structs. This was done simply to preserve
+the interface to an existing function (`CreateAndAddPrmDsc`) which
+processes each kernel caller parameter as a capture object member. 
```

Comment:
I actually thought that this is done to be able to pass the array into the kernel, as the top level capture struct is broken down into components, and the wrapping becomes single, and it needs to be at least single as bare (unwrapped) arrays can't be passed in C++ as mentied above.
So if the reason is different and *if* it is possible to create AST for an argument of bare (unwrapped) array type then you might want to consider not  wrapping the array and introducing a new kind of parameter - kind_array - instead of using kind_std_layout. That would probably simplify code generation and generated code. Maybe a TODO for future.

Reply:
The generated code uses a struct to wrap the array. The double-wrapping is an implementation detail, just the way some existing functions are used to achieve an array wrapped in a (single) struct as the generated code.

Reply:
We should fix the CreateAndAddParmDsc instead, a fixme here won't be fixable without being an ABI break

Reply:
There isn't a way in clang to represent a function argument that is a whole array. That's why a wrap in a struct is necessary.